src/main/java/com/ldtteam/structurize/items/ItemTagTool.java

```diff
@@ -199,29 +195,23 @@ public boolean canAttackBlock(final BlockState state, final Level worldIn, final
         // add/remove tags
         Map<BlockPos, List<String>> tagPosMap = ((IBlueprintDataProvider) te).getPositionedTags();
 
-        if (!tagPosMap.containsKey(relativePos))
+        if (!tagPosMap.containsKey(relativePos) || !tagPosMap.get(relativePos).contains(currentTag))
         {
             ((IBlueprintDataProvider) te).addTag(relativePos, currentTag);
-            Network.getNetwork().sendToServer(new AddRemoveTagMessage(true, currentTag, relativePos, anchorPos));
-
-            LanguageHandler.sendPlayerMessage(player, "com.ldtteam.structurize.gui.tagtool.addtag", currentTag, new TranslatableComponent(
-              worldIn.getBlockState(pos).getBlock().getDescriptionId()));
-        }
-        else if (!tagPosMap.get(relativePos).contains(currentTag))
-        {
-            ((IBlueprintDataProvider) te).addTag(relativePos, currentTag);
-            Network.getNetwork().sendToServer(new AddRemoveTagMessage(true, currentTag, relativePos, anchorPos));
-
-            LanguageHandler.sendPlayerMessage(player, "com.ldtteam.structurize.gui.tagtool.addtag", currentTag, new TranslatableComponent(
-              worldIn.getBlockState(pos).getBlock().getDescriptionId()));
+            if (worldIn.isClientSide())
+            {
+                LanguageHandler.sendPlayerMessage(player, "com.ldtteam.structurize.gui.tagtool.addtag", currentTag, new TranslatableComponent(
+                  worldIn.getBlockState(pos).getBlock().getDescriptionId()));
+            }
         }
         else
         {
             ((IBlueprintDataProvider) te).removeTag(relativePos, currentTag);
-            Network.getNetwork().sendToServer(new AddRemoveTagMessage(false, currentTag, relativePos, anchorPos));
```

Comment:
AddRemoveTagMessage do we still need this message now?

Reply:
Yes, we use it on remove in the UI yet.